packages/php-wasm/compile/php/php_wasm.c

```diff
@@ -100,18 +103,41 @@ EMSCRIPTEN_KEEPALIVE FILE *wasm_popen(const char *cmd, const char *mode)
 			return 0;
 		}
 
+        int **descv = malloc(sizeof(int *) * 3);
+
+        int *stdin = malloc(sizeof(int) * 3);
+        int *stdout = malloc(sizeof(int) * 3);
+        int *stderr = malloc(sizeof(int) * 3);
+
+        stdin[0] = 0;
+        stdin[1] = stdin_childend;
+        stdin[2] = stdin_parentend;
+
+        stdout[0] = 1;
+        stdout[1] = stdout_pipe[0];
+        stdout[2] = stdout_pipe[1];
+
+        stderr[0] = 2;
+        stderr[1] = stderr_pipe[0];
+        stderr[2] = stderr_pipe[1];
+
+        descv[0] = stdin;
+        descv[1] = stdout;
+        descv[2] = stderr;
+
 		// the wasm way {{{
 		js_open_process(
-			cmd, 
-			stdin_childend,
-			// stdout. @TODO: Pipe to /dev/null
-			stdout_pipe[0], 
-			stdout_pipe[1],
-			// stderr. @TODO: Pipe to /dev/null
-			stderr_pipe[0],
-			stderr_pipe[1]
+			cmd,
+            descv,
```

Comment:
Nice! TIL you can pass arrays like this

Reply:
Thinking about this more, this seems to add quite a few lines of code. What do you think about preserving the long `js_open_process` signature and, whenever a descriptor is missing, passing NULLs directly as function args?

Reply:
Yes! What is passed is the array pointer, and I had to loop over the memory to find all the data.

Reply:
A-ha! I missed the fact PHP actually uses those additional descriptors. Indeed this is [what the documentation says](https://www.php.net/proc_open):

> The file descriptor numbers are not limited to 0, 1 and 2 - you may specify any valid file descriptor number and it will be passed to the child process. This allows your script to interoperate with other scripts that run as "co-processes". In particular, this is useful for passing passphrases to programs like PGP, GPG and openssl in a more secure manner. It is also useful for reading status information provided by those programs on auxiliary file descriptors.

This is really a good spot, thank you! Let's use the array, then. I just checked, and this can actually be handled by Playground as Node.js `child_process.spawn()` [supports arbitrary file descriptors via the `stdio` option](https://nodejs.org/api/child_process.html#optionsstdio).